### DISCUSSION TITLE
Support for ESP32 Pico authorized GPIO

### DISCUSSION THREAD

**Discussion by hallard** [arendst/Tasmota#11756] [Ideas]
Dear ESP's guru

I made a board using latest ESP32-PICO-V3-02 which has in single chip ESP32 of course, 8Mb flash and 2mb PSRAM (this one is just awesome) letting my board size to be as small as 3cmx2cm (even with pcb antenna restrict).

It will be used for French Teleinfo smart metering with tasmota of course and here a picture for curious.

![image](https://user-images.githubusercontent.com/2471931/115084609-761bfd80-9f09-11eb-8323-dad0e208a579.png)

The trick is that the Pico D4 has onboard flash and different pinout than classic ESP32 module, here the table below, as you can see some I/O are available and other not, depending on module.

![image](https://user-images.githubusercontent.com/2471931/115088402-60f69d00-9f10-11eb-96aa-30bdfe5d6c4e.png)

When flashed tasmota, of course it works fine. But, I have the Serial input connected on GPIO8 (which I can use) but I can't select in tasmota. See below they are not on the list

![image](https://user-images.githubusercontent.com/2471931/115088786-43760300-9f11-11eb-8581-0944de80a065.png)

I can make a PR for that, it's no problem, but just wondering where is the "link" between board type and available pins on the GUI ?

Any direction by @arendst or @Jason2866 for that would be very welcome.

**Comment by s-hadinger** [comment]
Awesome. A new Denky. The form factor is impressively small. It really has 2mb of psram in the chip?

**Reply by hallard** [reply]
Yes, it's brand new chip, really awesome, we just need to wait this supported by IDF for this chip now :-)

**Reply by arendst** [reply]
The pico falls in the same league as the "normal" esp32. Nothing fancy apparently so no need for a different idf. Just learned when I introduced the new GetDeviceHardware() function.

All I need to provide is a working template with Pico GPIO's...

**Reply by hallard** [reply]
Correct, but this one has PSRAM wired different (GPIO) than usual external. So I'm not sure where to place:

- Enable PSRAM
- Define PSRAM GPIO

**Comment by arendst** [comment]
@hallard See my comment here https://github.com/arendst/Tasmota/pull/11757#issuecomment-821779942

**Comment by arendst** [comment]
Just looked over the GPIO differences between ESP32, ESP32-PICO-D4 and ESP32-PICO-V3 and they are a mess.

For the time being I suggest not to use GPIO6,7,8,9,10,11 (normal ESP32 flash connection), GPIO6,7,8,11,16,17 (Pico-d4 flash connection) and GPIO6,11,16,17,18,23 (Pico-v3 flash connection).

The S2 makes things even worse as flash is connected to GPIO27,28,29,30,31,32.

Tasmota only supports GPIOs 0 to 5,9,10,12 to 39. Implementing support for pins 6,7,8 and 11 is a nightmare.

So in short, just do not plan to use pins 6,7,8 and 11 in any project.

**Reply by kozak-olexandr** [reply]
> Just looked over the GPIO differences between ESP32, ESP32-PICO-D4 and ESP32-PICO-V3 and they are a mess.
> 
> For the time being I suggest not to use GPIO6,7,8,9,10,11 (normal ESP32 flash connection), GPIO6,7,8,11,16,17 (Pico-d4 flash connection) and GPIO6,11,16,17,18,23 (Pico-v3 flash connection).
> 
> The S2 makes things even worse as flash is connected to GPIO27,28,29,30,31,32.
> 
> Tasmota only supports GPIOs 0 to 5,9,10,12 to 39. Implementing support for pins 6,7,8 and 11 is a nightmare.
> 
> So in short, just do not plan to use pins 6,7,8 and 11 in any project.

Hi there!
I will add a little about ESP32-Pico D4. Now I'm testing controllers based on this processor. They work great with ESPHome. I want to try to flash Tasmota. The instructions for the controllers contain a GPIO plate that is used.
[https://res2.weblium.site/res/60376cae5c621400217c0c5a/60bf745982d9e700220617e0](url)

**Comment by arendst** [comment]
NB. As GPIO8 won't be available for user selection you can always hardcode this GPIO in your teleinfo project so the board layout is not a waste ;-)

**Reply by hallard** [reply]
Hummm i already made a fix but not PR waiting for dynamic template but if Gpio8 will never be available in it I’ll PR the fix

**Reply by hallard** [reply]
What about a compilation define allowing to expose all gpio in ui ? And this makes user responsible of what they doing with GpIo ?

**Comment by s-hadinger** [comment]
@hallard Is the board now available? Independently, I would like to test Tasmota with ESP32-PICO-V3-02 but I can't find any dev board yet.

**Reply by hallard** [reply]
@s-hadinger it's no problem, just sent you a discord PM

You need to accept me on discord, can't send you message.

**Comment by hallard** [comment]
@s-hadinger did you get a chance to add me on discord, or other channel as you wish?

**Reply by Jason2866** [reply]
@hallard  have you joined Tasmota Discord https://discord.gg/Ks2Kzd4 ? If no please do. I will give you the user right to talk to the staff. Ping me (Jason2866 on Discord too) that i know your Discord name.

**Reply by hallard** [reply]
@Jason2866 thanks, that's done

**Comment by s-hadinger** [comment]
With #13069 you can now configure GPIO8

**Reply by s-hadinger** [reply]
The template is now

```
{"NAME":"Denky D4","GPIO":[32,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,450,449,448,0,0,5632,0,0,0,0,0,0,0,0,1],"FLAG":0,"BASE":1}
```

But I need to fix something with serial on ESP32, Arduino on idf4.4 rewrote the uart driver.

**Reply by hallard** [reply]
Awesome, absolutely need to give it a try on spare time, thank you very much for this fix (and others)

**Reply by hallard** [reply]
@s-hadinger, @Jason2866 
Long time without tasmota, now I got some spare time and wanted to get back on Denky D4.
I see now some `platformio_tasmota_env32.ini` interesting files.
What advice you would give to use `ESP32-PICO-V3-02` to have 2M PSRAM and full 8M flash?

Until there I was using 2 different way for Denky D4, one with classic tasmota IDF config and the other with IDF 4.4 as follow

```ini
[env:denkyd4]
extends                     = env:tasmota32
board                       = pico32
platform_packages           = ${common32.platform_packages}
build_unflags               = ${common32.build_unflags}
build_flags                 = ${common32.build_flags}
                             -DUSE_TELEINFO
                             -DARDUINO_DENKY_PICOV3
;                             -DBOARD_HAS_PSRAM -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw


; *** EXPERIMENTAL Tasmota version for ESP32 IDF4.4.
[env:denkyd4-idf4]
extends                     = env:tasmota32_base
;board                       = pico32
board                       = esp32_8M
platform                    = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages           = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/375/framework-arduinoespressif32-master-32a9e33b1.tar.gz
                              platformio/tool-mklittlefs @ ~1.203.200522
build_unflags               = ${esp32_defaults.build_unflags}
                              -Wswitch-unreachable
                              -Wincompatible-pointer-types

build_flags                 = ${esp32_defaults.build_flags}
                              -Wno-switch-unreachable
                             -DFIRMWARE_TASMOTA32
                             -DUSE_TELEINFO
                             -DARDUINO_DENKY_PICOV3
                             -DBOARD_HAS_PSRAM -mfix-esp32-psram-cache-issue -mfix-esp32-psram-cache-strategy=memw
                             ;-DESP32_STAGE=true        
```

Since I don't see specific config for `ESP32-PICO-V3-02` I was wondering what should I do best, I'm pretty sure my config is outdated.

Thanks for any advice. After I'll test #13069 , very interested by this one.

**Comment by Jason2866** [comment]
@hallard Tasmota does not use individual boards definitions, since it is not needed.
Tasmota checks during startup which ESP32 variant it is and enables PSRAM when found and can be used, without the PSRAM fixes linked (only needed for old ESP32 revisions). So your setup
```
[env:denkyd4-idf4]
extends                     = env:tasmota32_base
board                       = esp32_8M
platform                    = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages           = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/375/framework-arduinoespressif32-master-32a9e33b1.tar.gz
                              platformio/tool-mklittlefs @ ~1.203.200522
build_unflags               = ${esp32_defaults.build_unflags}
                              -Wswitch-unreachable
                              -Wincompatible-pointer-types

build_flags                 = ${esp32_defaults.build_flags}
                             -Wno-switch-unreachable
                             -DFIRMWARE_TASMOTA32
                             -DUSE_TELEINFO
                             -DARDUINO_DENKY_PICOV3
                             ;-DESP32_STAGE=true
```
is nearly correct. I just removed the not needed PSRAM settings. It is in boards.json definition. The PSRAM fix libs are not needed for the Pico

**Reply by Jason2866** [reply]
You can try my latest ESP32 framework. You should get around 40k more free RAM.
Replace the one regarding line in the setup with this

```
platform_packages           = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/495/framework-arduinoespressif32-master-dc1041d15.tar.gz
```

**Reply by s-hadinger** [reply]
Indeed it should work with the standard settings. Just add TELEINFO flag

**Comment by hallard** [comment]
Thank you guys, done some tests this morning, according @s-hadinger, basic configuration for Denky D4 works

```ini
[env:denkyd4]
extends                 = env:tasmota32_8M
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_TASMOTA32
                        -DUSE_TELEINFO
```

Then decided to add BLE with `FIRMWARE_BLUETOOTH` since got plenty of sensors around but keeping teleinfo

```ini
[env:denkyd4-ble]
extends                 = env:tasmota32_8M
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_BLUETOOTH
                        -DUSE_TELEINFO
lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_ssl, 
                          lib/lib_div/LibTeleinfo
```

Does works either, so all is fine, except I don't see PSRAM but this is a normal behavior since it's not based on IDF4, correct?

Now tried @Jason2866 beta build (to get PSRAM working) with 

```ini
[env:denkyd4-idf4]
extends                 = env:tasmota32_8M
platform                = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages       = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/495/framework-arduinoespressif32-master-dc1041d15.tar.gz
                          platformio/tool-mklittlefs @ ~1.203.200522
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_TASMOTA32
                          -Wno-switch-unreachable
                          -DUSE_TELEINFO
build_unflags           = ${env:tasmota32_base.build_unflags}
                          -Wswitch-unreachable
                          -Wincompatible-pointer-types
```

Compiling is fine but it's unable to link, and looks like the error message is so helpfull (bad value)

```
Compiling /Users/charles/piobuild/denkyd4-idf4/FrameworkArduino/wiring_pulse.c.o
Compiling /Users/charles/piobuild/denkyd4-idf4/FrameworkArduino/wiring_shift.c.o
Archiving /Users/charles/piobuild/denkyd4-idf4/libFrameworkArduino.a
Indexing /Users/charles/piobuild/denkyd4-idf4/libFrameworkArduino.a
Linking /Users/charles/piobuild/denkyd4-idf4/firmware.elf
/Users/charles/.platformio/packages/toolchain-xtensa32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
*** [/Users/charles/piobuild/denkyd4-idf4/firmware.elf] Error 1
============================================================ [FAILED] Took 93.75 seconds ============================================================

Environment    Status    Duration
-------------  --------  ------------
denkyd4-idf4   FAILED    00:01:33.751
======================================================= 1 failed, 0 succeeded in 00:01:33.751 =======================================================

```

Just saw new release today, I also tried with master of tamota (with latest ESP32 Core 2.0.1 rc1) and tried to build default `tasmota32idf4`

```ini
; *** Beta Tasmota version for ESP32 IDF4.4.
[env:tasmota32idf4]
extends                     = env:tasmota32_base
platform                    = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages           = framework-arduinoespressif32 @ https://github.com/tasmota/arduino-esp32/releases/download/2.0.1rc1/framework-arduinoespressif32-release_IDF4.4.tar.gz
                              platformio/tool-mklittlefs @ ~1.203.200522
build_unflags               = ${env:tasmota32_base.build_unflags}
                              -Wswitch-unreachable
                              -Wstringop-overflow
                              -Wincompatible-pointer-types
build_flags                 = ${env:tasmota32_base.build_flags}
                              -Wno-switch-unreachable
                              -Wno-stringop-overflow
; ***Uncomment next two lines to enable BLE
                              ;-DUSE_BLE_ESP32
                              ;-DUSE_MI_ESP32
                              ;-DESP32_STAGE=true
```

Looks like no more success on link stage, not sure what I'm doing wrong with IDF4 beta?
```
Compiling /Users/charles/piobuild/tasmota32idf4/FrameworkArduino/wiring_pulse.c.o
Compiling /Users/charles/piobuild/tasmota32idf4/FrameworkArduino/wiring_shift.c.o
Archiving /Users/charles/piobuild/tasmota32idf4/libFrameworkArduino.a
Indexing /Users/charles/piobuild/tasmota32idf4/libFrameworkArduino.a
Archiving /Users/charles/piobuild/tasmota32idf4/lib61b/lib_mail.a
Indexing /Users/charles/piobuild/tasmota32idf4/lib61b/lib_mail.a
Linking /Users/charles/piobuild/tasmota32idf4/firmware.elf
/Users/charles/.platformio/packages/toolchain-xtensa32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
*** [/Users/charles/piobuild/tasmota32idf4/firmware.elf] Error 1
============================================================ [FAILED] Took 93.17 seconds ============================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32idf4  FAILED    00:01:33.175
======================================================= 1 failed, 0 succeeded in 00:01:33.175 =======================================================
```

**Comment by Jason2866** [comment]
Which esptool version do you have or better is used in your attempts?
You need esptool.py v.3.1 for successfully linking.
To be sure add this to `platform_packages`
`platformio/tool-esptoolpy @ ~1.30100`

**Reply by hallard** [reply]
Same thing, not surprised, note that my issue is not on upload stage but on the link stage, was not aware esptool was used also on link stage, good to know.

are you able to build the default target named `tasmota32idf4`?

**Comment by Jason2866** [comment]
It does build:
```
Indexing .pio/build/denkyd4-idf4/libFrameworkArduino.a
Linking .pio/build/denkyd4-idf4/firmware.elf
Retrieving maximum program size .pio/build/denkyd4-idf4/firmware.elf
Checking size .pio/build/denkyd4-idf4/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  18.3% (used 59804 bytes from 327680 bytes)
Flash: [=====     ]  46.4% (used 1428233 bytes from 3080192 bytes)
Building .pio/build/denkyd4-idf4/firmware.bin
esptool.py v3.1
Merged 25 ELF sections
bin_map_copy([".pio/build/denkyd4-idf4/firmware.bin"], [".pio/build/denkyd4-idf4/firmware.elf"])
map_gzip([".pio/build/denkyd4-idf4/firmware.bin"], [".pio/build/denkyd4-idf4/firmware.elf"])
====================================== [SUCCESS] Took 66.09 seconds ======================================

Environment    Status    Duration
-------------  --------  ------------
denkyd4-idf4   SUCCESS   00:01:06.086
====================================== 1 succeeded in 00:01:06.086 ======================================
```
Used this
```
[env:denkyd4-idf4]
extends                 = env:tasmota32_8M
platform                = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages       = framework-arduinoespressif32 @ https://github.com/tasmota/arduino-esp32/releases/download/2.0.1rc1/framework-arduinoespressif32-release_IDF4.4.tar.gz
                          platformio/tool-mklittlefs @ ~1.203.200522
build_unflags           = ${env:tasmota32_base.build_unflags}
                          -Wswitch-unreachable
                          -Wstringop-overflow
                          -Wincompatible-pointer-types
build_flags             = ${env:tasmota32_base.build_flags}
                          -DFIRMWARE_TASMOTA32
                          -Wno-switch-unreachable
                          -Wno-stringop-overflow
                          -DUSE_TELEINFO

```
So your Platformio setup is messed up. Quick fix is to delete hidden folder `.platformio`. No fear Platformio does completly rebuild.

**Comment by hallard** [comment]
Insane, did it, and got PSRAM seen thanks. 

Now I need to fix MQTT TLS not working anymore :-)

**Comment by s-hadinger** [comment]
Great. Does teleinfo work? I had some issues with earlier version of idf4.4. I hope it is solved now.

**Comment by hallard** [comment]
> Great. Does teleinfo work? I had some issues with earlier version of idf4.4. I hope it is solved now.

Not tested all in detail, but Legacy mode is working, I'm still investigating on MQTTs not connecting with valid let's Encrypt certificate, I'm writing a detailed issue I suspect you're the MQTT TLS's guru

**Comment by Jason2866** [comment]
Hello Charles,
since today no more special setup needed. Tasmota32 uses now Arduino ESP32 based on IDF4.4 by default.
